### PR TITLE
demote to java 17, since helpmojo plugin does not support java 18+

### PR DESCRIPTION
**Pull request by ericvergnaud**

### CODE REVIEW

**Comment by kaby76** [comment]
What is HelpMojo and why is it needed?

**Review by KvanTTT — COMMENTED** [review]
Agree with @kaby76. I want to figure out what is HelpMojo and why it's important. Could you please explain reasoning?

The argument against: upgrading version can significantly affect our future user-base.

**Comment by ericvergnaud** [comment]
In the end state the tool will be compiled to WebAssembly (so we can also run it from the web). So users won't be able to use a Java version that doesn't support WebAssembly, and the latest version of GraalVM (which doesn't yet support GC) seems to require version 21 (see https://www.graalvm.org/latest/reference-manual/wasm/).
So it's good practice to start upgrading our tech stack asap, before we have actual users and have to deal with backwards compatibility.
I tried to upgrade to 21 directly but running a fresh build raised the HelpMojo issue, so I had to move back to java 17.
HelpMojo is a goal of the [maven-plugin-plugin](https://maven.apache.org/plugin-tools/maven-plugin-plugin/) used for the build of the antlr-maven-plugin.

**Comment by kaby76** [comment]
> HelpMojo is a goal of the [maven-plugin-plugin](https://maven.apache.org/plugin-tools/maven-plugin-plugin/) used for the build of the antlr-maven-plugin.

Are you planning on separating the build for the Antlr Tool from the the antlr-maven-plugin? Seems like a plugin has nothing to do with building the Antlr Tool, although for testing you need some build rule support?

**Comment by KvanTTT** [comment]
> I tried to upgrade to 21 directly but running a fresh build raised the HelpMojo issue

What issue have you encountered? Maybe at first it makes sense try to fix this issue?

> so I had to move back to java 17.

Does it make sense to upgrade to 17 considering WA is only compatible with JDK >= 21 (as you also said)?

File: runtime-testsuite/test/org/antlr/v5/test/runtime/java/api/perf/TimeLexerSpeed.java [hunk 1]
```diff
@@ -166,8 +166,8 @@
  *  @since 4.7
  */
 public class TimeLexerSpeed { // don't call it Test else it'll run during "mvn test"
-	public static final String Parser_java_file = "Java/src/org/antlr/v4/runtime/Parser.java";
-	public static final String RuleContext_java_file = "Java/src/org/antlr/v4/runtime/RuleContext.java";
+	public static final String Parser_java_file = "Java/src/org/antlr/v5/runtime/Parser.java";
+	public static final String RuleContext_java_file = "Java/src/org/antlr/v5/runtime/RuleContext.java";
```
Comment (KvanTTT) [inline runtime-testsuite/test/org/antlr/v5/test/runtime/java/api/perf/TimeLexerSpeed.java:170]: Please don't mix unrelated changes in the same PR. Tool version upgrade doesn't relate to JDK version upgrade.

File: runtime-testsuite/pom.xml [hunk 2]
```diff
@@ -66,7 +66,7 @@
 		<dependency>
 		    <groupId>org.openjdk.jol</groupId>
 		    <artifactId>jol-core</artifactId>
-		    <version>0.16</version>
+		    <version>0.17</version>
```
Comment (KvanTTT) [inline runtime-testsuite/pom.xml]: Is this updrade needed? `0.16` isn't compatible with with JDK 17, am I right? If so, please put these details into the commit message.

File: .gitignore [hunk 3]
```diff
@@ -137,3 +137,5 @@ runtime/Cpp/runtime/libantlr4-runtime.dylib
 
 # Go test and performance trace files
 **/*.pprof
+runtime-testsuite/gen
+tool-testsuite/gen
```
Comment (KvanTTT) [inline .gitignore:141]: Please extract unrelated changes to a separated PR.

**Review by KvanTTT — CHANGES_REQUESTED** [review]

File: runtime-testsuite/pom.xml (hunk 2 above)
Comment (ericvergnaud) [reply runtime-testsuite/pom.xml]: I think this one is a typo

**Comment by ericvergnaud** [comment]
What doesn't work is that HelpMojo cannot open .class files with 65 version number or above

**Comment by ericvergnaud** [comment]
It is good practice to progress towards the target even if we can't yet match it. It helps detect and address issues earlier. This PR is a perfect example of that i.e. we know we need to address the HelpMojo issue in order to use Java > 17.

**Comment by ericvergnaud** [comment]
@KvanTTT you probably want to approve this so we can merge it and have a working Java CI as part of #6

**Comment by KvanTTT** [comment]
> @KvanTTT you probably want to approve this so we can merge it and have a working Java CI as part of https://github.com/antlr/antlr5/pull/6

I can approve it but it looks like CI from different branches should not affect each other because they have their own CI configuration files (`hoted.yml`).

**Review by KvanTTT — APPROVED** [review]

**Comment by ericvergnaud** [comment]
CI is running now for this PR
